proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,521 @@
+      TextField("Enter the item's name...", $shoppingItem.name)
+      
+    }
+  }
+  
+}
+```
+
+
+## Proposed solution
+
+We propose to extend the contexts where application of property-wrapper types is permitted. Namely, application of such types will be allowed on function and closure parameters:
+
+
+## Detailed design
+
+Property wrappers are essentially sugar wrapping a given property with compiler synthesized code. This proposal retains this principle employing the following rules for transformation.
+
+### Property wrappers on function parameters
+
+A function parameter marked with property-wrapper custom attributes must conform to the following rules:
+
+1. Property wrapper function parameters must support initialization through their `wrappedValue` type. Therefore, all property-wrapper types must provide a suitable `init(wrappedValue:)`.
+2. Each `wrappedValue` getter must be `nonmutating`.
+3. Default values for such parameters must be expressed in terms of the innermost `wrappedValue` type.
+
+Transformation of property-wrapper parameters will be performed as such:
+
+1. The argument label will remain unchanged.
+2. The parameter name will be prefixed with an underscore.
+3. The parameter will be bound to the backing property-wrapper type.
+4. A local computed property representing  `wrappedValue` will be synthesized by the compiler and named per the original – non-prefixed – parameter name. Moreover, the accessors will mirror the `wrappedValue` accessors. Lastly, a setter will _only_ be synthesized for the local property if the `wrappedValue`'s setter is `nonmutating`, or if the wrapper is a reference type.
+5. If the property wrapper defines a `projectedValue` property, a local computed property representing `projectedValue` will be synthesized by the compiler and named per the original parameter name prefixed with a dollar sign (`$`). The same accessor rules for `wrappedValue` apply to `projectedValue`.
+6. When passing an argument to a property wrapper parameter, the compiler will wrap the argument in the appropriate `init(wrappedValue:)` call.
+
+#### Transformation example:
+
+```swift
+@propertyWrapper
+struct Lowercased {
+
+  init(wrappedValue: String) { ... }
+    
+    
+  var wrappedValue: String {
+    get { ... }
+    set { ... }
+  }
+  
+}
+
+func postUrl(
+  @Lowercased urlString: String
+) { ... }
+
+
+postUrl(urlString: "mySite.xyz/myUnformattedUsErNAme")
+```
+
+In the above code, the `postUrl(urlString:)` function and its caller are equivalent to:
+
+```swift
+func postUrl(urlString _urlString: Lowercased) {
+
+  var urlString: String {
+    get { _ urlString.wrappedValue }
+    // The setter accessor is not synthesized
+    // because the setter of the 'wrappedValue' 
+    // property of 'Lowercased' is mutating.
+  }
+
+
+  ...
+  
+}
+
+
+postUrl(urlString: Lowercased(wrappedValue: "mySite.xyz/myUnformattedUsErNAme"))
+```
+
+#### Restrictions on property wrapper function parameters
+
+##### `@autoclosure`
```

Comment:
This part seems to belong to __Alternatives considered__; _not_ __Detailed design__. I propose we simple say that `@autoclosure` is not supported, maybe note that this is discussed in more detailed at the  __Alternatives considered__ section, and place it in that section.

Reply:
This is a justification for why `@autoclosure` is not supported. It's not really an alternative. We could explicitly label this as **Justification** or **Rationale**; I saw that in one of the [concurrency proposals](https://github.com/DougGregor/swift-evolution/blob/async-await/proposals/nnnn-async-await.md#asynchronous-functions) and I liked that the justification for a design decision was made explicit.

Reply:
I think this _really_ looks like an Alternatives Considered topic; __however__, if you think it should be included in the Detailed Design section as a sort of Rationale, I think that's okey. We do, though, – at least in my opinion – have to make it slightly smaller and in more of a note style. So, for example, if we shortened it keeping the most important parts, and put it in a note style as is done in the concurrency proposal, think that would be better.

#### Just to see how it would look like 


>*__RATIONALE:__* Function parameters with a property-wrapper custom-attribute cannot have an `@autoclosure` type, because `@autoclosure` is _unnecessary_ for the wrapped value itself. That is, the wrapped-value argument at the call-site will _always_ be wrapped in a call to `init(wrappedValue:)`, which can already support `@autoclosure` arguments. Furthermore, using `@autoclosure` for the backing wrapper-type would be misleading, since the type spelled out for the parameter would not be the true autoclosure type; consider the `reportProgress` example from above with an `@autoclosure` parameter:
>
 > ```swift
> func postUrl(
>  @Lowercased urlString: @autoclosure () -> String
> ) { ... }
>
>
> postUrl(urlString: "mySite.xyz/myUnformattedUsErNAme")
> ```
 > The above code would be transformed to:
 > ```swift
> func postUrl(
>   urlString _urlString: @autoclosure () -> Lowercased
> ) { ... }
>
>
> postUrl(urlString: Lowercased(wrappedValue: "mySite.xyz/myUnformattedUsErNAme"))
> ```
 > As you can see, the changing type of the `@autoclosure` is incredibly misleading, as it's not obvious that `@autoclosure` applies to the backing wrapper rather than the wrapped value. For these reasons, using `@autoclosure` for a property-wrapper function-parameter will be invalid.

Reply:
Totally agree that the rationale could be smaller. How about we explain the first part about using `@autoclosure` for the wrapped value type here:

Function parameters with a property-wrapper custom attribute cannot have an @autoclosure type.

> RATIONALE: `@autoclosure` is unnecessary for the wrapped value itself. That is, the wrapped-value argument at the call-site will always be wrapped in a call to `init(wrappedValue:)`, which can already support `@autoclosure` arguments.

And the rest, which is really talking about supporting `@autoclosure` for the backing wrapper, can go in Alternatives Considered. What do you think?

Reply:
I think it turned out very good! As for any further explanation that, you mentioned, could be moved in Alternatives Considered, I don't think it's necessary; what's already in Detailed Design is already sufficient – in my opinion.